### PR TITLE
Introduce Waitable.

### PR DESCRIPTION
**Pull request by iocanel**
We have discussed those changes as part of #613

### CODE REVIEW

**Comment by iocanel** [comment]
@jimmidyson: since you have some spare, I thought it might be a good idea to keep you off boredom :-)

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/internal/readiness/Readiness.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    return spec.getReplicas().intValue() == status.getReadyReplicas();
+  }
+
+
+  public static boolean isDeploymentConfigReady(DeploymentConfig d) {
+    Utils.checkNotNull(d, "Deployment can't be null.");
+    DeploymentConfigSpec spec = d.getSpec();
+    DeploymentConfigStatus status = d.getStatus();
+
+    return spec.getReplicas().intValue() == status.getReplicas() &&
```
Comment (jimmidyson) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/internal/readiness/Readiness.java]: `status.getReadyReplicas()`?

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/internal/readiness/Readiness.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+      throw new IllegalArgumentException("Item needs to be one of [Deployment, ReplicaSet, Pod, DeploymentConfig, ReplicationController], but was: [" + item.getKind() + "]");
+    }
+  }
+
+  public static boolean isDeploymentReady(Deployment d) {
+    Utils.checkNotNull(d, "Deployment can't be null.");
+    DeploymentSpec spec = d.getSpec();
+    DeploymentStatus status = d.getStatus();
+
+    return spec.getReplicas().intValue() == status.getReplicas() &&
```
Comment (jimmidyson) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/internal/readiness/Readiness.java]: `status.getReadyReplicas()`?

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/internal/readiness/Readiness.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+      spec.getReplicas().intValue() <= status.getAvailableReplicas();
+  }
+
+  public static boolean isReplicationControllerReady(ReplicationController r) {
+    Utils.checkNotNull(r, "ReplicationController can't be null.");
+    ReplicationControllerSpec spec = r.getSpec();
+    ReplicationControllerStatus status = r.getStatus();
+
+    return spec.getReplicas().intValue() == status.getReplicas();
```
Comment (jimmidyson) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/internal/readiness/Readiness.java]: `status.getReadyReplicas()`?

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/internal/readiness/Readiness.java (hunk 2 above)
Comment (iocanel) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/internal/readiness/Readiness.java]: not available.

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/internal/readiness/Readiness.java (hunk 1 above)
Comment (iocanel) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/internal/readiness/Readiness.java]: Not available.

Comment (jimmidyson) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/internal/readiness/Readiness.java]: Sorry just noticed that's only available in OpenShift 1.5 :(

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/internal/readiness/Readiness.java (hunk 2 above)
Comment (jimmidyson) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/internal/readiness/Readiness.java]: Sorry just noticed that's only available in Kubernetes master :(

**Comment by jimmidyson** [comment]
[merge]

**Comment by palmerabollo** [comment]
Do you know why it is not implemented for Service Accounts? Is it a bug or a feature? Thanks!

**Comment by rohanKanojia** [comment]
@palmerabollo : Could you please elaborate your use case? Does it make sense for ServiceAccounts to be Waitable?

**Comment by palmerabollo** [comment]
Hi @rohanKanojia, THANK YOU for your prompt reply. As far as I know, when you create a Service Account in Kubernetes, there is another process (the Token Controller, see https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/#token-controller) that is in charge of observing Service Account creations + creating the Secret to allow API access.

I've faced some cases where the Service Account is created but the token/secret is still not associated to it, so the Service Account is not yet usable. This is why I think it should be `Waitable`, to be able to wait for the token/secret to be ready, too. Does it make sense?

Currently I have to implment a custom polling mechanism to get the Service Account status and wait until the secret is associated to the Service Account before continuing. 

I might be missing something or perhaps there is a cleaner workaround I'm not aware of.

**Pull request merged**